sys/linux/vusb.txt

```diff
@@ -0,0 +1,687 @@
+include <uapi/linux/usb/cdc.h>
+
+resource fd_usb[fd]
+
+syz_usb_connect(speed flags[usb_device_speed], dev_len len[dev], dev ptr[in, usb_device_descriptor], dev_id ptr[in, usb_device_id], conn_descs ptr[in, vusb_connect_descriptors]) fd_usb
+syz_usb_control_io(fd fd_usb, descs ptr[in, vusb_descriptors], resps ptr[in, vusb_responses])
+syz_usb_ep_write(fd fd_usb, ep int16[0:31], len len[data], data buffer[in])
+syz_usb_disconnect(fd fd_usb)
+
+usb_device_speed = USB_SPEED_UNKNOWN, USB_SPEED_LOW, USB_SPEED_FULL, USB_SPEED_HIGH, USB_SPEED_WIRELESS, USB_SPEED_SUPER, USB_SPEED_SUPER_PLUS
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# Special struct generated by Go code.
+usb_device_id {
+	idVendor	int16
+	idProduct	int16
+	bcdDevice	int16
+	bDeviceClass	int8
+	bDeviceSubClass	int8
+	bDeviceProtocol	int8
+	bIntClass	int8
+	bIntSubClass	int8
+	bIntProtocol	int8
+	bIntNumber	int8
+} [packed]
+
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+# # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # # #
+
+# USB device, configuration, interface and endpoint descriptors.
+
+# TODO: support more than one configuration.
+usb_device_descriptor {
+	bLength			const[USB_DT_DEVICE_SIZE, int8]
+	bDescriptorType		const[USB_DT_DEVICE, int8]
+
+	bcdUSB			int16
+	bDeviceClass_p		flags[usb_classes, int8]
+	bDeviceSubClass_p	int8
+	bDeviceProtocol_p	int8
+	bMaxPacketSize0		int8
+	idVendor_p		int16
+	idProduct_p		int16
+	bcdDevice_p		int16
+	iManufacturer		int8
+	iProduct		int8
+	iSerialNumber		int8
+	bNumConfigurations	len[configs, int8]
+
+	configs			array[usb_config_descriptor, 1]
+} [packed]
+
+usb_classes = USB_CLASS_PER_INTERFACE, USB_CLASS_AUDIO, USB_CLASS_COMM, USB_CLASS_HID, USB_CLASS_PHYSICAL, USB_CLASS_STILL_IMAGE, USB_CLASS_PRINTER, USB_CLASS_MASS_STORAGE, USB_CLASS_HUB, USB_CLASS_CDC_DATA, USB_CLASS_CSCID, USB_CLASS_CONTENT_SEC, USB_CLASS_VIDEO, USB_CLASS_WIRELESS_CONTROLLER, USB_CLASS_MISC, USB_CLASS_APP_SPEC, USB_CLASS_VENDOR_SPEC
+
+# TODO: support more than one interface.
+usb_config_descriptor {
+	bLength			const[USB_DT_CONFIG_SIZE, int8]
+	bDescriptorType		const[USB_DT_CONFIG, int8]
+
+	wTotalLength		len[parent, int16]
+	bNumInterfaces		len[interfaces, int8]
+	bConfigurationValue	int8
+	iConfiguration		int8
+	bmAttributes		flags[usb_config_attributes, int8]
+	bMaxPower		int8
+
+	interfaces		array[usb_interface_descriptor, 1]
+} [packed]
+
+usb_config_attributes = USB_CONFIG_ATT_ONE, USB_CONFIG_ATT_SELFPOWER, USB_CONFIG_ATT_WAKEUP, USB_CONFIG_ATT_BATTERY
+
+usb_interface_descriptor {
+	bLength			const[USB_DT_INTERFACE_SIZE, int8]
+	bDescriptorType		const[USB_DT_INTERFACE, int8]
+
+	bInterfaceNumber_p	int8
```

Comment:
Please name the fields as they are named in kernel sources. This field does not have _p suffix.
Soon we will hopefully do static checking and name matching will be required for meaningful checking.

Reply:
Done.